### PR TITLE
Exports framework draft

### PR DESCRIPTION
**Pull request by foxbunny**
This patch lays down a framework for handling component member exports (routes, hooks, plugins, databases, assets, etc). With this framework:

- all exports are defined in components' configuration files, in `[exports]` section
- all exports are clearly visible in the configuration files so they do not have to be hunted down in different places in the code base (good for reviewing 3rd party apps, for example)
- exports can live in arbitrary python modules
- no need to mess with the master configuration file (the exports collectors manage their own state without touching the master configuration)
- ability to define new kinds of exports _in external components_
- better documentation (yay!)

The configuration format is documented in `docs/source/developers/component_exports.rst`.

This is still not finished. In particular:

- Tutorial for writing your own collectors is missing
- Unit tests for concrete collectors (in `librarian.core.collectors` package) are missing
- There are still missing collectors (see #273)
- Everything is generally untested

I propose that we merge this asap, so that other cleanup tasks can start taking advantage of it as soon as possible.

### CODE REVIEW

File: docs/source/developers/components_exports.rst [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+- hooks
+- plugins
+- WSGI middleware
+- route handlers
+- command line argument handlers
+
+Each component may have one or more members that belong to one of the CMGs. The
+CMGs are also known as component member type. 
+
+All members from all components that belong to the same group comprise that
```
Comment (goodbadwolf) [inline docs/source/developers/components_exports.rst]: `All members from all components that belong to the same group comprise that group` is a bit hard to parse.

File: docs/source/developers/components_exports.rst [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+a template named 'foo', and component B also has that template, the last
+component's template will always take precendece and override any previously
+found templates of the same name. Also, keep in mind that overriding templates
+in this way *intentionally* is also a bad idea, as the override order is an
+implementation detail that may change in future without notice.
+
+Component configuration file
+----------------------------
+
+Each component has a configuration file. This confugration file is in .ini
```
Comment (goodbadwolf) [inline docs/source/developers/components_exports.rst]: Typo `confugration`

File: docs/source/developers/components_exports.rst [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+    The exact format used in the configuration files is covered `in confloader
+    module documentation
+    <http://confloader.readthedocs.org/en/latest/writing_ini.html>`_.
+
+Object and path references
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When specifying relative paths in the configuration options, you cannot use
+double-dot notation anywhere. For example, ``static_dir = ../foo`` is not
+allowed. The path resolver code will automatically strip out the ``../`` part.
```
Comment (goodbadwolf) [inline docs/source/developers/components_exports.rst]: This should clarify that the paths used are relative to the component directory and not the config's directory

File: docs/source/developers/components_exports.rst [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+Templates
+---------
+
+By default, templates used in a component are looked for in a directory called
+'views', located inside the component's package directory. This can be changed
+by specifying one or more alternative locations. For example::
+
+    [exports]
+
+    templats = templates
```
Comment (goodbadwolf) [inline docs/source/developers/components_exports.rst]: Typo `templats`

File: docs/source/developers/components_exports.rst [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+
+It is important to remember that templates are resolved relative to template
+driectories, regardless of how many leves of hierarchy there is between the
```
Comment (goodbadwolf) [inline docs/source/developers/components_exports.rst]: Typo `driectories`

File: docs/source/developers/components_exports.rst [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+                foo1.tpl
+                foo2.tpl
+            bar/
+                baz/
+                    baz1.tpl
+                    baz2.tpl
+                bar1.tpl
+                bar2.tpl
+
+If we used the second configuration from the examples above, and we ask for a
```
Comment (goodbadwolf) [inline docs/source/developers/components_exports.rst]: There is one example of config and one example of folder hierarchy under the Templates section, so it's unclear which is referring to.

File: docs/source/developers/components_exports.rst [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+        article: article_layout
+        summary: summary_layout
+
+This confiuration creates two JavaScript bundles, and two CSS bundles. The
+JavaScript bundle 'article' contains concatenated and minified sources of
+'static/js/autoscroll.js' and 'static/js/comment.js'. As you can see, the
+static assets directory and 'js/' subdirectory are automatically added to the
+name, as is the '.js' extension.
+
+For more information on working with templates, see
```
Comment (goodbadwolf) [inline docs/source/developers/components_exports.rst]: This should be `For more information on working with static assets, ...`

File: docs/source/developers/components_exports.rst [hunk 8]
```diff
@@ -0,0 +1,494 @@
...
+
+    hooks = hooks.on_initialize
+
+In this case, we have an ``on_initialize`` function in a ``hooks`` module.
+
+Plugins
+-------
+
+Plugins are classes and functions that follow the `Bottle's plugin API
+<http://bottlepy.org/docs/dev/plugindev.html>`_. Just like middleware, plugins
```
Comment (goodbadwolf) [inline docs/source/developers/components_exports.rst]: Not sure if it makes a difference, but ideally this should link to the stable version of the doc at [Plugin API](http://bottlepy.org/docs/stable/plugindev.html)

File: docs/source/developers/components_exports.rst [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+
+.. note::
+    Dependency/dependents declaration is completely optional. It is only needed
+    if order matters. If you are reasonably sure that it does not matter where
+    in the stack your plugin is executed, you may skip to the end of this
+    subsection.
+
+Plugins can declare dependencies on each other using
+:py:func:`~librarian.core.exports.depends_on` and
+:py:func:`~librarian.core.exports.requried_by` decorators. For semantic
```
Comment (goodbadwolf) [inline docs/source/developers/components_exports.rst]: Typo `requried_by`here and a few places below

Comment (foxbunny) [inline docs/source/developers/components_exports.rst]: Bottle generally keeps backwards compatibility in each release, so this is ok.

File: docs/source/developers/components_exports.rst [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+:py:func:`~librarian.core.exports.requried_by` decorators, respectively.
+
+Here are a few examples::
+
+    import functools
+
+    from librarian.core.exports import *
+
+    
+    @depend_on('foo')
```
Comment (goodbadwolf) [inline docs/source/developers/components_exports.rst]: Typo `depend_on`

File: docs/source/developers/components_exports.rst [hunk 11]
```diff
@@ -0,0 +1,494 @@
...
+    @before(['bar', 'baz'])
+    def my_other_plugin(handler):
+        ....
+    my_other_plugin.name = 'myfantasticplugin'
+
+
+    class MyPlugin(object):
+        name = 'myexcuisiteplugin'
+        api = 2
+        dependencies = ['foo', 'bar']
```
Comment (goodbadwolf) [inline docs/source/developers/components_exports.rst]: Are these aliases to depends_on, required_by ?

Comment (foxbunny) [inline docs/source/developers/components_exports.rst]: `@before` and `@after` are aliases for `@depensd_on` and `@required_by`.

Comment (goodbadwolf) [inline docs/source/developers/components_exports.rst]: I meant that for `dependencies` and `dependents` used in the class example

Comment (foxbunny) [inline docs/source/developers/components_exports.rst]: Oh, I probably forgot to fox the documentation.

Comment (foxbunny) [inline docs/source/developers/components_exports.rst]: Fixed in new commit.

File: librarian/core/collectors/commands.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+                logging.error('Invalid handler {} for component {}'.format(
+                    command, component.name))
+                continue
+            handler.component = component.name
+            self.register(handler)
+
+    def install_member(self, handler):
+        name = handler.name
+        if name in self.handlers:
+            logging.warn('Duplicate registration for command: {}'.format(name))
```
Comment (goodbadwolf) [inline librarian/core/collectors/commands.py]: Can we print info about the existing handler ? That will help in debug/fixing

Comment (foxbunny) [inline librarian/core/collectors/commands.py]: Good call.

File: librarian/core/collectors/commands.py [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    def collect(self, component):
+        commands = component.get_export('commands', [])
+        for command in commands:
+            try:
+                handler = component.get_object(command)
+            except ImportError:
+                logging.exception('Could not import handler')
+                continue
+            if not hasattr(handler, 'name'):
+                logging.error('Invalid handler {} for component {}'.format(
```
Comment (goodbadwolf) [inline librarian/core/collectors/commands.py]: Can we have a more descriptive error message about what is invalid

Comment (foxbunny) [inline librarian/core/collectors/commands.py]: Fixed by not doing this check at all, and instead letting the install_member() call raise AttributeError which is logged as a traceback.

File: librarian/core/utils.py [hunk 14]
```diff
@@ -1 +1,208 @@
...
+    """
+    Iterator that can be mutated while being iterated over.
+
+    The main difference between :py:class:`muter` objects and other iterator
+    types offered by the Python's standard library is that these objects
+    support mutation during iteration. In particular, new elements can be
+    appended to the end of the iterator even when iterating over it. The
+    :py:meth:`~muter.append` method is used to append new items to the
+    iterator. While iterating, the code simply continues to iterate as long as
+    there are elements in the iterator, even if those elements were added afte
```
Comment (goodbadwolf) [inline librarian/core/utils.py]: Typo `afte`, missing `r`

File: docs/source/developers/components_exports.rst [hunk 15]
```diff
@@ -0,0 +1,508 @@
...
+<http://bottlepy.org/docs/dev/plugindev.html>`_. Just like middleware, plugins
+are applied in order like an onion skin. The last plugin that is registered is
+applied first, and the first plugin registered will be applied last. The
+request is intercepted by the last (outermost) plugin, and is passed through
+down the chain to the innermost plugin, which hands it over to the actual
+request handler. ::
+
+    plugins:        p1   p2   p3   p4   |
+                    |    |    |    |    |
+    reuqest  --->---+----+----+----+----+--\
```
Comment (goodbadwolf) [inline docs/source/developers/components_exports.rst]: Typo: reuqest

File: docs/source/developers/download_and_setup.rst [hunk 16]
```diff
@@ -34,8 +34,8 @@ Obtaining the sources
 
 Sources can be obtained by cloning from the `Librarian GitHub repository
 <https://github.com/Outernet-Project/librarian/>`_. (Also see
-:doc:`git_repository_layout`.) To clone the complete source tree from the mater
-branch::
+:doc:`../appendices/git_repository_layout`.) To clone the complete source tree
+from the mater branch::
```
Comment (goodbadwolf) [inline docs/source/developers/download_and_setup.rst]: Typo: mater

**Comment by integricho** [comment]
Merged along with feature/dashboard-menu-collectors
